src/librustdoc/html/render/cache.rs

```diff
@@ -623,13 +603,35 @@ fn build_index(krate: &clean::Crate, cache: &mut Cache) -> String {
         .map(|module| shorten(plain_summary_line(module.doc_value())))
         .unwrap_or(String::new());
 
+    let crate_aliases = aliases
+        .iter()
+        .map(|(k, values)| {
+            (
+                k.clone(),
+                values
+                    .iter()
+                    .filter_map(|v| {
+                        let x = &crate_items[*v];
+                        if x.parent_idx.is_some() == x.parent.is_some() { Some(*v) } else { None }
+                    })
+                    .collect::<Vec<_>>(),
+            )
+        })
+        .filter(|(_, values)| !values.is_empty())
+        .collect::<Vec<_>>();
+
     #[derive(Serialize)]
     struct CrateData<'a> {
         doc: String,
         #[serde(rename = "i")]
         items: Vec<&'a IndexItem>,
         #[serde(rename = "p")]
         paths: Vec<(ItemType, String)>,
+        // The String is alias name and the vec is the list of the elements with this alias.
+        //
+        // To be noted: the `usize` elements are indexes to `items`.
+        #[serde(rename = "a")]
+        aliases: Option<Vec<(String, Vec<usize>)>>,
```

Comment:
I don't want the field to be added if there are no alias, which is allowed through the `Option` type.

Reply:
> I don't want the field to be added if there are no alias, which is allowed through the `Option` type.

Using `Option` means the empty case is serialized as `a: null` rather than `a: []`. I don't see the advantage to that. If you don't want the field to be serialized at all then you can use [`#[serde(skip_serializing_if = "Vec::is_empty")]`](https://serde.rs/field-attrs.html#skip_serializing_if).



> What would be the gain of using a `BTreeMap` btw? Do we really care about being more deterministic?

We need determinism for reproducible builds. Admittedly it looks like `FxHashMap` is deterministic enough for that but I really don't like relying on that. Having it alphabetized also makes it easier to debug and compare changes made to the alias list.

Reply:
Let's go for a `BTreeMap` then!

Reply:
I also added `skip_serializing_if`.